Schema updates in database page without reloading. Fixes #2736

Aritra8438 opened the pull request:
**This PR Fixes** #2736 

<!-- Concisely describe what the pull request does. -->
**Technical details**
- As mentioned in the bug, the database page wasn't reflecting the recent schema count changes unless reloaded.
- This PR fixes the bug by updating the schema count on the addition or deletion of tables and explorations.
- Also to reduce the database call count, the schema updation request is only sent to the backend if actually addition or deletion happens, otherwise not.
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->
**Screencast**

https://user-images.githubusercontent.com/64671908/227736838-5da2eee1-02f0-4e61-9ea9-0f05578af45d.mp4


<!-- Add screenshots to show the problem and the solution; or delete the section entirely. -->


## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

Aritra8438 commented:
Hi, @pavish and @rajatvijay, here is my PR. Please suggest any improvements, I will be happy to acknowledge them. Thank you.

Aritra8438 reviewed (commented):
Hi @pavish, now the reloading logic will happen within the stores on the updation and deletion of tables and queries.

pavish reviewed (dismissed):
The `refetchSchema` method should not be called within components.

We should be calling them from the `tables` store within the `createTable` and `deleteTable` functions, similarly for explorations, it should happen within the `queries` store within `createQuery` and `deleteQuery` functions. It would be better if we modify the `num_tables` and `num_queries` directly instead of fetching the schemas all over again.

Aritra8438 commented:
Hi, @pavish. Sorry for the delay. **As, you have mentioned**, I
- moved the **schema updation logic to the respective stores**. 
- modified the code so that it **modifies the num_tables and num_queries directly** instead of fetching the schemas all over again.

Aritra8438 commented:
Hi, @pavish, a friendly ping, if you get time.

pavish reviewed (commented):
@Aritra8438 

The changes you made increased our reliance on `currentSchemaId` and `currentDbName` stores within the methods in `table.ts` and `queries.ts`. While, I can see why you followed this approach as we've had similar usages, we intend to refactor these stores to make them less complex and more readable and reducing inter dependencies.

I've added a couple commits rewriting some of the logic:
1. https://github.com/centerofci/mathesar/pull/2745/commits/b952650da928c41d8bd0c96d12eb6140f73a2cd0 passes down both database and schema when `createTable` and `deleteTable` methods are called. We always have this data with us, and passing these down already simplify usage within the tables.ts store, and would be more maintainable in the future.
2. https://github.com/centerofci/mathesar/pull/2745/commits/1a8d2a99ed2ae01640d2f95155ef04d8b7bc2c09 has changes to reduce usage of `currentSchemaId` and `currentDbName` within queries.ts. Unfortunately, the places where these methods are called do not contain the db and schema information. So, it's hard to pass them down without a major refactor. I've followed the less resistance path in this commit. Eventually, we would refactor this to be able to pass both database and schema information to these method.

@seancolsen I'm adding you as the reviewer to take a look at this again since I've rewritten most of the code.

Aritra8438 commented:
Thanks, @pavish.
Though it was mentioned that **Our store structures need a complete refactoring**, I didn't have enough understanding of the codebase to do so. So, I followed the similar approaches to solve the issue.
